# Cleaning the data

In [1]:
import pandas as pd
import numpy as np

Read in ratings and movies data

In [3]:
ratings_messy = pd.read_csv("IMDb ratings.csv")
films_messy = pd.read_csv("IMDb movies.csv")

/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3441: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


Set title IDs as index to smoothly join both tables, then drop unnecessary columns.

In [4]:
films_messy = films_messy.set_index('imdb_title_id')
ratings_messy = ratings_messy.set_index('imdb_title_id')

Check how many movies we can work with in this data, since metascore is our response variable.

In [6]:
len(films_messy['metascore'].dropna())

13305

Join both tables, then drop unnecessary columns.

In [7]:
combined_messy = pd.concat([films_messy, ratings_messy], axis=1)
combined_messy.columns

Index(['title', 'original_title', 'year', 'date_published', 'genre',
       'duration', 'country', 'language', 'director', 'writer',
       'production_company', 'actors', 'description', 'avg_vote', 'votes',
       'budget', 'usa_gross_income', 'worlwide_gross_income', 'metascore',
       'reviews_from_users', 'reviews_from_critics', 'weighted_average_vote',
       'total_votes', 'mean_vote', 'median_vote', 'votes_10', 'votes_9',
       'votes_8', 'votes_7', 'votes_6', 'votes_5', 'votes_4', 'votes_3',
       'votes_2', 'votes_1', 'allgenders_0age_avg_vote',
       'allgenders_0age_votes', 'allgenders_18age_avg_vote',
       'allgenders_18age_votes', 'allgenders_30age_avg_vote',
       'allgenders_30age_votes', 'allgenders_45age_avg_vote',
       'allgenders_45age_votes', 'males_allages_avg_vote',
       'males_allages_votes', 'males_0age_avg_vote', 'males_0age_votes',
       'males_18age_avg_vote', 'males_18age_votes', 'males_30age_avg_vote',
       'males_30age_votes', 'males_45age_av

In [21]:
ratings = combined_messy.drop(['original_title',
                               'date_published',
                               'duration',
                               'country',
                               'language',
                               'director',
                               'writer',
                               'production_company',
                               'actors',
                               'description',
                               'avg_vote',
                               'votes',
                               'budget',
                               'usa_gross_income',
                               'worlwide_gross_income',
                               'reviews_from_users',
                               'reviews_from_critics',
                               'allgenders_0age_avg_vote', # Dropping this because so many NAs
                               'allgenders_0age_votes',
                               'allgenders_18age_votes',
                               'allgenders_30age_votes',
                               'allgenders_45age_votes',
                               'males_allages_votes',
                               'males_0age_avg_vote',      # Dropping this because so many NAs
                               'males_0age_votes',
                               'males_18age_votes',
                               'males_30age_votes',
                               'males_45age_votes',
                               'females_allages_votes',
                               'females_0age_avg_vote',    # Dropping this because so many NAs
                               'females_0age_votes',
                               'females_18age_votes',
                               'females_30age_votes',
                               'females_45age_votes',
                               'top1000_voters_rating',
                               'top1000_voters_votes',
                               'us_voters_rating',
                               'us_voters_votes',
                               'non_us_voters_rating',
                               'non_us_voters_votes'],
                              1)

/var/folders/5r/j3mtn4sx1wv_xsdwft49lcdm0000gn/T/ipykernel_75831/2136986086.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  ratings = combined_messy.drop(['original_title',


As alluded to in the commented code, we planned to keep all of the average vote columns, but including the ones from critics under the age of 18 left a lot of missing data. This is presumably because it's not very common for people younger than 18 to be professional movie critics. The only column we want to lose data to is the metascore column, as that is our response.

In [22]:
ratings = ratings.dropna()

In [23]:
len(ratings)

13216

This looks good. We're only missing 89 movies that have a reported metascore to missing values from other columns.

In [24]:
ratings.head()

title  year                  genre  metascore  \
imdb_title_id                                                                
tt0006864              Intolerance  1916         Drama, History       99.0   
tt0017136               Metropolis  1927          Drama, Sci-Fi       98.0   
tt0018037      Il cantante di jazz  1927  Drama, Music, Musical       66.0   
tt0018773                 Il circo  1928        Comedy, Romance       90.0   
tt0019777            The Cocoanuts  1929        Comedy, Musical       69.0   

               weighted_average_vote  total_votes  mean_vote  median_vote  \
imdb_title_id                                                               
tt0006864                        7.8        13875        7.8          8.0   
tt0017136                        8.3       156076        8.1          8.0   
tt0018037                        6.5         8866        6.8          7.0   
tt0018773                        8.1        27414        7.9          8.0   
tt0019777                        7.0         6900        7.2          7.0   

               votes_10  votes_9  ...  allgenders_30age_avg_vote  \
imdb_title_id                     ...                              
tt0006864          3477     2230  ...                        7.7   
tt0017136         37520    36112  ...                        8.2   
tt0018037           999      589  ...                        6.4   
tt0018773          4542     5699  ...                        8.1   
tt0019777           813      496  ...                        6.9   

               allgenders_45age_avg_vote  males_allages_avg_vote  \
imdb_title_id                                                      
tt0006864                            7.8                     7.8   
tt0017136                            8.3                     8.3   
tt0018037                            6.6                     6.4   
tt0018773                            8.1                     8.1   
tt0019777                            7.1                     6.9   

               males_18age_avg_vote  males_30age_avg_vote  \
imdb_title_id                                               
tt0006864                       7.9                   7.7   
tt0017136                       8.3                   8.2   
tt0018037                       6.5                   6.4   
tt0018773                       8.2                   8.1   
tt0019777                       6.8                   6.8   

               males_45age_avg_vote  females_allages_avg_vote  \
imdb_title_id                                                   
tt0006864                       7.8                       7.5   
tt0017136                       8.3                       8.3   
tt0018037                       6.6                       6.6   
tt0018773                       8.1                       8.2   
tt0019777                       7.1                       7.2   

               females_18age_avg_vote  females_30age_avg_vote  \
imdb_title_id                                                   
tt0006864                         7.2                     7.4   
tt0017136                         8.2                     8.3   
tt0018037                         6.7                     6.4   
tt0018773                         8.3                     8.1   
tt0019777                         7.1                     7.1   

               females_45age_avg_vote  
imdb_title_id                          
tt0006864                         8.0  
tt0017136                         8.3  
tt0018037                         6.8  
tt0018773                         8.2  
tt0019777                         7.3  

[5 rows x 29 columns]

# Integrate genres into the clean dataframe

There are 13216 films in our dataset, each of which have one or more genres associated with them in the following format:  
  
tt0006864              Drama, History  
tt0017136               Drama, Sci-Fi  
tt0018037       Drama, Music, Musical  
tt0018773             Comedy, Romance  
tt0019777             Comedy, Musical  
                       ...             
tt9845398                       Drama  
tt9883996    Biography, Comedy, Drama  
tt9892094                       Drama  
tt9894470       Action, Crime, Horror  
tt9898858              Action, Comedy  
  
We see that some films have multiple genres in the same string, so we have some cleaning to do. We need to figure out a way to find all the unique genres and create columns for each one, indicating which genre is associated with a given film using a 1 (True) or 0 (False).

Start by creating a list of the genres for each movie.

The following code iterates through the list of movie's genres and, if there are multiple, splits them into lists. If we come across a genre that we haven't seen yet, we add it to the list ```genres_unique```.

In [13]:
genres = ratings.genre

In [14]:
genres_unique = []

for film in range(0,len(genres)):
    film_genres = genres[film].split(", ")
    for genre in film_genres:
        if (genre not in genres_unique):
            genres_unique.append(genre)

Let's check if the list is indeed unique:

In [16]:
#pd.Series(genres_unique).nunique()
# 21
#len(genres_unique)
# 21

21

There are 21 unique genres associated with the films in the dataset. Now we will cycle through each genre, creating a column for each one and assigning 1 if the film contains that genre, and 0 if it doesn't.

In [17]:
for genre in genres_unique:
    ratings[genre] = ratings['genre'].str.contains(genre).astype(int)

# Drop the messy genres column
ratings = ratings.drop('genre', axis=1)

The dataset now includes genre data in one-hot encoded form.

In [18]:
ratings.head()

title  year  metascore  weighted_average_vote  \
imdb_title_id                                                                
tt0006864              Intolerance  1916       99.0                    7.8   
tt0017136               Metropolis  1927       98.0                    8.3   
tt0018037      Il cantante di jazz  1927       66.0                    6.5   
tt0018773                 Il circo  1928       90.0                    8.1   
tt0019777            The Cocoanuts  1929       69.0                    7.0   

               total_votes  mean_vote  median_vote  votes_10  votes_9  \
imdb_title_id                                                           
tt0006864            13875        7.8          8.0      3477     2230   
tt0017136           156076        8.1          8.0     37520    36112   
tt0018037             8866        6.8          7.0       999      589   
tt0018773            27414        7.9          8.0      4542     5699   
tt0019777             6900        7.2          7.0       813      496   

               votes_8  ...  Adventure  Family  Fantasy  Mystery  Thriller  \
imdb_title_id           ...                                                  
tt0006864         3214  ...          0       0        0        0         0   
tt0017136        39299  ...          0       0        0        0         0   
tt0018037         1135  ...          0       0        0        0         0   
tt0018773         9607  ...          0       0        0        0         0   
tt0019777         1362  ...          0       0        0        0         0   

               Biography  Animation  Action  Film-Noir  Sport  
imdb_title_id                                                  
tt0006864              0          0       0          0      0  
tt0017136              0          0       0          0      0  
tt0018037              0          0       0          0      0  
tt0018773              0          0       0          0      0  
tt0019777              0          0       0          0      0  

[5 rows x 50 columns]

Export data:

In [ ]:
ratings.to_csv('ratings_clean.csv')

In [ ]:
ratings.columns

In [ ]:
films_messy[['title']]